### PR TITLE
add proper validation to room edit and creation

### PR DESCRIPTION
**Pull request by denisdimitrov**
Bugs fixed:
- You can no longer create a room without logo.
- You can no longer edit a room and leave empty name or url.
- You can no longer edit a room without a logo.
- Newly created rooms are now enabled by deafult.
- Room stays enabled after edit

Fixes: https://github.com/sparkletown/internal-sparkle-issues/issues/374
Fixes: https://github.com/sparkletown/internal-sparkle-issues/issues/375

### CODE REVIEW

**Comment by 0xdevalias** [comment]
> Newly created rooms are now enabled by default

@denisdimitrov Might want to check with @sparkletown/non-tech-team about that, and/or look at the code history that made it that way. I believe that may have been a deliberate choice that we made to have them not enabled by default.

File: src/pages/Admin/Details/ValidationSchema.ts
Comment (0xdevalias) [inline src/pages/Admin/Details/ValidationSchema.ts:36]: Why not using `lowerCamelCase` for `image_url`?

File: src/pages/Admin/Details/ValidationSchema.ts
Comment (0xdevalias) [inline src/pages/Admin/Details/ValidationSchema.ts:175]: Is there a reason why we can't reuse the same schema stuff that is used on creation? What differs between the two? And why does it differ? Feels like they should both validate with the same code?

File: src/pages/Admin/Room/Edit/RoomEdit.tsx
Comment (0xdevalias) [inline src/pages/Admin/Room/Edit/RoomEdit.tsx:39]: As mentioned elsewhere RE: using the same validation as `roomCreateSchema`

File: src/pages/Admin/Room/Edit/RoomEdit.tsx
Comment (0xdevalias) [inline src/pages/Admin/Room/Edit/RoomEdit.tsx:44]: This seems like a bug/typo?
```suggestion
      template: room.template,
```

File: src/pages/Admin/Room/Modal/Item/Item.tsx
Comment (0xdevalias) [inline src/pages/Admin/Room/Modal/Item/Item.tsx:67]: I'm assuming this is the part I called out in https://github.com/sparkletown/sparkle/pull/1176#issuecomment-779632114 that required further checking?

File: src/pages/Admin/Room/Modal/Item/Item.tsx
Comment (0xdevalias) [inline src/pages/Admin/Room/Modal/Item/Item.tsx:102]: What does the `false` here do? It's hard to tell given we aren't using named parameters on `setValue`

**Review by 0xdevalias — COMMENTED** [review]
Questions/comments

File: src/pages/Admin/Details/ValidationSchema.ts
Comment (denisdimitrov) [reply src/pages/Admin/Details/ValidationSchema.ts:36]: This is legacy code in the firebase, if I change it here. I'll have to change it in another 20-30 places. I don't think this should be done in this PR.

File: src/pages/Admin/Details/ValidationSchema.ts
Comment (denisdimitrov) [reply src/pages/Admin/Details/ValidationSchema.ts:175]: When creating a room there is an option to select either new venue and room or simple room with url. That validation fails in  the edit because that field doesn't exist anywhere in the firebase.

File: src/pages/Admin/Details/ValidationSchema.ts
Comment (0xdevalias) [reply src/pages/Admin/Details/ValidationSchema.ts:36]: Yup, agreed OOS for here. Just wanted to know the 'why'

File: src/pages/Admin/Room/Modal/Item/Item.tsx
Comment (denisdimitrov) [reply src/pages/Admin/Room/Modal/Item/Item.tsx:67]: Correct.

File: src/pages/Admin/Room/Modal/Item/Item.tsx
Comment (denisdimitrov) [reply src/pages/Admin/Room/Modal/Item/Item.tsx:102]: The third parameter was `triggerValidation` and it was set on `false`. And when you upload a new image the validation was not triggering.

File: src/pages/Admin/Details/ValidationSchema.ts
Comment (0xdevalias) [reply src/pages/Admin/Details/ValidationSchema.ts:175]: Fair enough. I'm not deeply familiar with yup, but is there a way we can share the same partial validation rules between the 2? It looks like it's just an object, so perhaps we can move the object outside of the `.shape()` call, and then use the common bits from that to build the 2 separate validations.

File: src/pages/Admin/Room/Modal/Item/Item.tsx
Comment (0xdevalias) [reply src/pages/Admin/Room/Modal/Item/Item.tsx:67]: Cool. Probably worth leaving this open until we have a resolution one way or the other.

File: src/pages/Admin/Room/Modal/Item/Item.tsx
Comment (0xdevalias) [reply src/pages/Admin/Room/Modal/Item/Item.tsx:102]: Makes sense.

File: src/pages/Admin/Room/Edit/RoomEdit.tsx
Comment (denisdimitrov) [reply src/pages/Admin/Room/Edit/RoomEdit.tsx:39]: https://github.com/sparkletown/sparkle/pull/1176#discussion_r576619116

**Comment by denisdimitrov** [comment]
@0xdevalias 
> @denisdimitrov Might want to check with @sparkletown/non-tech-team about that, and/or look at the code history that made it that way. I believe that may have been a deliberate choice that we made to have them not enabled by default.

I just got verification from Ed that rooms are supposed to be enabled. That's how it is in the legacy code.

File: src/pages/Admin/Details/ValidationSchema.ts
Comment (denisdimitrov) [reply src/pages/Admin/Details/ValidationSchema.ts]: I created some constants that are being shared between them.

File: src/pages/Admin/Room/Modal/Item/Item.tsx
Comment (denisdimitrov) [reply src/pages/Admin/Room/Modal/Item/Item.tsx:67]: https://github.com/sparkletown/sparkle/pull/1176#issuecomment-779703122

**Comment by jess-sparkles** [comment]
To answer the above question, having rooms enabled right after creating is great! We asked for this back in December so thanks for fixing it!

**Comment by 0xdevalias** [comment]
> We asked for this back in December so thanks for fixing it!

@jess-sparkles Interesting. I don't believe that ever flowed through to me (or I forgot about it since if it did). Good to know that this will solve that for you though :)

Is that a universal want/need? Or is it something that it would be useful to have a toggle option on this form (that defaults to show)? (as my memory was that we made them hidden by default by explicit request, as per https://github.com/sparkletown/sparkle/pull/1176#issuecomment-779632114)

**Comment by jess-sparkles** [comment]
Seems like on Slack @Sparkle-T wants them to be turned off by default (so no change). @sofisparkle and I wanted them turned on by default as we didn't think it was obvious that the rooms were turned off until you go back to the admin page (after making the room) and we thought this would be confusing for users.

**Comment by 0xdevalias** [comment]
> Seems like on Slack @Sparkle-T wants them to be turned off by default (so no change). @sofisparkle and I wanted them turned on by default as we didn't think it was obvious that the rooms were turned off until you go back to the admin page (after making the room) and we thought this would be confusing for users.

@jess-sparkles By my read of his @Sparkle-T's [comments on slack](https://doesitsparkle.slack.com/archives/C01B7MSMA3W/p1613554268235800?thread_ts=1613524973.230900&cid=C01B7MSMA3W) it sounded like he was agreeing with you both, that there is no reason to have them default to off since most events are created/set up in advance (though noted that having the ability to configure that in future might be a nice feature, for those 'edge cases' where someone may want to change the default)

File: src/pages/Admin/Details/ValidationSchema.ts
Comment (0xdevalias) [inline src/pages/Admin/Details/ValidationSchema.ts:173]: In `roomCreateSchema`, this is only present within the `.when("useUrl")`. Is that logic also required here?

**Review by 0xdevalias — APPROVED** [review]
[CRO] One small question/possible refinement, but aside from that I think this looks good to go (assuming it's been properly clicked through/tested/etc)

---

**Edit:** Actually @denisdimitrov, 1 more question, I never saw a reply to https://github.com/sparkletown/internal-sparkle-issues/issues/375#issuecomment-778905257 confirming whether it is even actually a bug, or if it's 'feature as desired/expected'. Would be good to confirm this before merging:

> @0xdevalias:
> I'm not 100% sure that this is a bug per-se. Didn't we want the ability to be able to not have to explicitly upload an image on all of these sorts of things, and if we didn't, that we would use a default image instead? (cc // @sparkleed @sparkletown/product)

File: src/pages/Admin/Details/ValidationSchema.ts
Comment (denisdimitrov) [reply src/pages/Admin/Details/ValidationSchema.ts:173]: It is not because that field is being used only to determine if it should validate url or venue name, when editing a room you are always validating the url.

Comment (0xdevalias) [reply src/pages/Admin/Details/ValidationSchema.ts:173]: Cool. That works for me!

**Pull request merged**
